src/main/java/com/artipie/asto/s3/InternalExceptionHandle.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 artipie.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto.s3;
+
+import com.artipie.asto.FailedCompletionStage;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * Translate an exception happened inside future.
+ *
+ * @param <T> Future result type.
+ * @since 0.34
+ */
+final class InternalExceptionHandle<T> implements BiFunction<T, Throwable, CompletionStage<T>> {
+
+    /**
+     * The original future.
+     */
+    private final CompletableFuture<T> future;
+
+    /**
+     * Type of exception to handle.
+     */
+    private final Class<? extends Throwable> from;
+
+    /**
+     * Converter to a new exception.
+     */
+    private final Function<? super Throwable, ? extends Throwable> convert;
+
+    /**
+     * Ctor.
+     *
+     * @param future Original future.
+     * @param from Internal type of exception.
+     * @param convert Converter to a external type.
+     */
+    InternalExceptionHandle(
+        final CompletableFuture<T> future,
```

Comment:
@andreoss why do we need to depend on `future` here? `thenApply` method takes optional result and error and returns the result. In case of matched exception `this.from.isInstance(throwable.getCause())` we can throw converted exception from method body:
```java
    @Override
    public CompletionStage<T> apply(final T content, final Throwable throwable) {
        CompletionStage<T> result = this.future;
        if (
            throwable instanceof CompletionException
                && this.from.isInstance(throwable.getCause())
        ) {
            throw new FailedCompletionStage<>(
                this.convert.apply(throwable.getCause())
            );
        }
        return content;
    }
```

The problem with current code is that it adds implicit dependency on `future` and potenticaly may cause some side effects or unexpected behavior (e.g. if we cancel the future in this method or if we inject another future by the mistake)

Reply:
@g4s8 Done.